File: oc/Quartz2DDemo/Quartz2DDemo/ViewController.m
//
//  ViewController.m
//  Quartz2DDemo
//
//  Created by dfang on 2018-1-26.
//  Copyright © 2018年 east. All rights reserved.
//

#import "ViewController.h"
#import "CustomView.h"
#import "ProgressCircle.h"
#import "HuangView.h"
#import "PieView.h"
#import "BalloonView.h"
#import "WaterImageView.h"
#import "GraffitiView.h"
#import "LockView.h"

@interface ViewController () <LockViewDelegate>

@property (nonatomic, strong) UISlider *slider;
@property (nonatomic, strong) ProgressCircle *progressCircle;

@property (nonatomic, strong) UIImageView *imageView;

@property (nonatomic, strong) GraffitiView *graffiti;
@property (nonatomic, strong) UIButton *colorBtn;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    // CustomView
//    CustomView *customView = [[CustomView alloc] initWithFrame:self.view.frame];
//    customView.backgroundColor = [UIColor whiteColor];
//    [self.view addSubview:customView];
    
    // ProgressCircle
//    [self.view addSubview:self.progressCircle];
//    [self.view addSubview:self.slider];
    
    // HuangView
//    HuangView *huang = [[HuangView alloc] initWithFrame:self.view.frame];
//    huang.backgroundColor = [UIColor whiteColor];
//    [self.view addSubview:huang];
    
    // PieView
//    PieView *pie = [[PieView alloc] initWithFrame:self.view.frame];
//    pie.backgroundColor = [UIColor whiteColor];
//    pie.sections = @[@20, @50, @30];
//    pie.sectionColors = @[[UIColor yellowColor], [UIColor greenColor], [UIColor blueColor]];
//    [self.view addSubview:pie];
    
    // BalloonView
//    BalloonView *balloon = [[BalloonView alloc] initWithFrame:self.view.frame];
//    balloon.backgroundColor = [UIColor lightGrayColor];
//    [self.view addSubview:balloon];

    // WaterImageView
//    _imageView = [[UIImageView alloc] initWithFrame:self.view.frame];
//    _imageView.image = [UIImage imageNamed:@"scene"];
//    [self.view addSubview:_imageView];
    
    // GraffitiView
//    [self testGraffitiView];
    
    // LockView
    LockView *lock = [[LockView alloc] initWithFrame:self.view.frame];
    lock.backgroundColor = [UIColor colorWithPatternImage:[UIImage imageNamed:@"gesture_bg"]];
    lock.delegate = self;
    lock.center = self.view.center;
    [self.view addSubview:lock];
}

// test for ProgressCircle
- (UISlider *)slider {
    if (_slider == nil) {
        _slider = [[UISlider alloc] initWithFrame:CGRectMake(50, 200, 200, 30)];
        [_slider addTarget:self action:@selector(sliderChanged:) forControlEvents:UIControlEventValueChanged];
    }
    return _slider;
}

- (ProgressCircle *)progressCircle {
    if (_progressCircle == nil) {
        _progressCircle = [[ProgressCircle alloc] initWithFrame:CGRectMake(50, 50, 240, 128)];
        _progressCircle.backgroundColor = [UIColor whiteColor];
    }
    return  _progressCircle;
}

- (void)sliderChanged:(UISlider *)sender {
    self.progressCircle.progress = sender.value;
}

// test WaterImageView
- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    UIImage *newImage = [WaterImageView waterImageWithBgImageName:@"scene" waterImageName:@"logo" scale:2];
    self.imageView.image = newImage;
}

// test GraffitiView
- (void)testGraffitiView {
    CGFloat padding = 30;
    CGFloat marginTop = 50;
    CGFloat btnWidth = (self.view.frame.size.width - padding * 4) / 3;
    CGFloat btnHeight = 40;
    self.graffiti = [[GraffitiView alloc] initWithFrame:CGRectMake(10, marginTop + btnHeight + 20, self.view.frame.size.width - 20, 350)];
    
    UIButton *backBtn = [[UIButton alloc] initWithFrame:CGRectMake(padding, marginTop, btnWidth, btnHeight)];
    backBtn.backgroundColor = [UIColor cyanColor];
    [backBtn setTitle:@"返回" forState:UIControlStateNormal];
    [backBtn addTarget:self.graffiti action:@selector(backClicked) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:backBtn];
    
    UIButton *clearBtn = [[UIButton alloc] initWithFrame:CGRectMake(padding + (btnWidth + padding), marginTop, btnWidth, btnHeight)];
    clearBtn.backgroundColor = [UIColor cyanColor];
    [clearBtn setTitle:@"清除" forState:UIControlStateNormal];
    [clearBtn addTarget:self.graffiti action:@selector(clearClicked) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:clearBtn];
    
    UIButton *saveBtn = [[UIButton alloc] initWithFrame:CGRectMake(padding + (btnWidth + padding) * 2, marginTop, btnWidth, btnHeight)];
    saveBtn.backgroundColor = [UIColor cyanColor];
    [saveBtn setTitle:@"保存" forState:UIControlStateNormal];
    [saveBtn addTarget:self.graffiti action:@selector(saveClicked) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:saveBtn];
    
    self.graffiti.backgroundColor = [UIColor grayColor];
    [self.view addSubview:self.graffiti];
    
    self.colorBtn = [[UIButton alloc] initWithFrame:CGRectMake((self.view.frame.size.width - btnWidth * 2)/2, CGRectGetMaxY(self.graffiti.frame) + 10, btnWidth * 2, btnHeight)];
    self.colorBtn.backgroundColor = [UIColor blackColor];
    [self.colorBtn setTitle:@"随机颜色" forState:UIControlStateNormal];
    [self.colorBtn addTarget:self action:@selector(colorBtnClicked) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:self.colorBtn];
}

    // test for GraffitiView
- (void)colorBtnClicked {
    UIColor *color = [UIColor colorWithRed:arc4random_uniform(256)/255.0 green:arc4random_uniform(256)/255.0 blue:arc4random_uniform(256)/255.0 alpha:1.0];
    self.graffiti.currentColor = color;
    self.colorBtn.backgroundColor = color;
}

    // LockViewDelegate
- (void)unlockWithPassword:(NSString *)password {
    NSLog(@"password:%@", password);
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: swift/Quartz2DDemo/Quartz2DDemo/ViewController.swift
//
//  ViewController.swift
//  Quartz2DDemo
//
//  Created by dfang on 2018-2-2.
//  Copyright © 2018年 east. All rights reserved.
//

import UIKit

class ViewController: UIViewController, LockViewProtocol {
    
    let progressCircle = ProgressCircleView.init()
    let waterImageView = UIImageView.init()
    let graffitiView = GraffitiView.init()
    let colorBtn = UIButton.init(type: UIButtonType.system)
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        self.view.backgroundColor = UIColor.white
        // CustomView
//        let customView = CustomView.init(frame: self.view.frame)
//        customView.backgroundColor = UIColor.clear
//        self.view.addSubview(customView)
        
        // HuangView
//        let huang = HuangView.init(frame: self.view.frame)
//        huang.backgroundColor = UIColor.clear
//        self.view.addSubview(huang)
        
        // PieView
//        let pie = PieView.init(frame: self.view.frame)
//        pie.backgroundColor = UIColor.clear
//        pie.sections = [20, 30, 10, 40, 20]
//        pie.sectionColors = [UIColor.red, UIColor.green, UIColor.blue, UIColor.brown, UIColor.cyan]
//        self.view.addSubview(pie)
        
        // ProgressCircleView
//       progressCircle.frame = CGRect.init(origin: CGPoint.zero, size: CGSize.init(width: self.view.frame.size.width, height: self.view.frame.size.height - 100))
//        progressCircle.backgroundColor = UIColor.clear
//        self.view.addSubview(progressCircle)
//        let slider = UISlider.init(frame: CGRect.init(x: 20, y: self.view.frame.size.height - 80, width: self.view.frame.size.width - 40, height: 30))
//        slider.minimumValue = 0
//        slider.maximumValue = 1
//        slider.setValue(0, animated: true)
//        slider.addTarget(self, action: #selector(sliderValueChanged(slider:)), for: UIControlEvents.valueChanged)
//        self.view.addSubview(slider)
        
        // SnowView
//        let snow = SnowView.init(frame: self.view.frame)
//        snow.backgroundColor = UIColor.clear
//        self.view.addSubview(snow)
        
        // WaterImageView
//        waterImageView.frame = self.view.frame
//        waterImageView.image = UIImage.init(named: "papa")
//        self.view.addSubview(waterImageView)

        // ScreenshotsView
//        let screenshots = ScreenshotsView.init(frame: self.view.frame)
//        screenshots.backgroundColor = UIColor.blue
//        self.view.addSubview(screenshots)
        
        // GraffitiView
//        self.testGraffitiView()
        
        // LockView
        let lock = LockView.init(frame: self.view.frame)
        lock.delegete = self
        // 设置背景图片
        lock.backgroundColor = UIColor.init(patternImage: UIImage.init(named: "gesture_bg")!)
        self.view.addSubview(lock)
    }
    
    // test for ProgressCircleView
    @objc func sliderValueChanged(slider: UISlider) {
        progressCircle.progress = slider.value
        progressCircle.setNeedsDisplay()
    }
    
    // test for WaterImageView
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        if (waterImageView.image == UIImage.init(named: "papa")) {
            let waterImage = WaterImageView.waterImageWith(bgImageName: "papa", waterImageName: "snow", scale: 0.8)
            waterImageView.image = waterImage
        } else {
            waterImageView.image = UIImage.init(named: "papa")
        }
    }
    
    func testGraffitiView() {
        let size = self.view.frame.size
        let marginLeft:CGFloat = 20
        let marginTop:CGFloat = 50
        let btnWidth = (size.width - marginLeft * 4) / 3
        let btnHeight:CGFloat = 50
        
        let backBtn = UIButton.init(type: UIButtonType.system)
        backBtn.frame = CGRect.init(x: marginLeft, y: marginTop, width: btnWidth, height: btnHeight)
        backBtn.setTitle("返回", for: UIControlState.normal)
        backBtn.backgroundColor = UIColor.cyan
        backBtn.addTarget(graffitiView, action: #selector(graffitiView.back), for: UIControlEvents.touchUpInside)
        self.view.addSubview(backBtn)
        
        let clearBtn = UIButton.init(type: UIButtonType.system)
        clearBtn.frame = CGRect.init(x: marginLeft + (btnWidth + marginLeft), y: marginTop, width: btnWidth, height: btnHeight)
        clearBtn.setTitle("清除", for: UIControlState.normal)
        clearBtn.backgroundColor = UIColor.cyan
        clearBtn.addTarget(graffitiView, action: #selector(graffitiView.clear), for: UIControlEvents.touchUpInside)
        self.view.addSubview(clearBtn)
        
        let saveBtn = UIButton.init(type: UIButtonType.system)
        saveBtn.frame = CGRect.init(x: marginLeft + (btnWidth + marginLeft) * 2, y: marginTop, width: btnWidth, height: btnHeight)
        saveBtn.setTitle("保存", for: UIControlState.normal)
        saveBtn.backgroundColor = UIColor.cyan
        saveBtn.addTarget(graffitiView, action: Selector(("saveToAlbum")), for: UIControlEvents.touchUpInside)
        self.view.addSubview(saveBtn)
        
        graffitiView.frame = CGRect.init(x: 10, y: marginTop + btnHeight + 30, width: size.width - 20, height: 280)
        graffitiView.backgroundColor = UIColor.gray
        self.view.addSubview(graffitiView)
        
        colorBtn.frame = CGRect.init(x: (size.width - 100)/2, y: size.height - 80, width: 100, height: 50)
        colorBtn.setTitle("随机更改颜色", for: UIControlState.normal)
        colorBtn.backgroundColor = UIColor.black
        colorBtn.addTarget(self, action: #selector(colorBtnClicked), for: UIControlEvents.touchUpInside)
        self.view.addSubview(colorBtn)
    }
    
    // test for GraffitiView
    @objc func colorBtnClicked() {
        let r = CGFloat(arc4random() % 256) / 255.0
        let g = CGFloat(arc4random() % 256) / 255.0
        let b = CGFloat(arc4random() % 256) / 255.0
        let color = UIColor.init(red: r, green: g, blue: b, alpha: 1.0)
        graffitiView.curLineColor = color
        colorBtn.backgroundColor = color
    }

    // test for LockView
    func unlockWithPassword(password: String) {
        print("password:" + password)
    }
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}

